File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.cpp
#include<string>
#include <queue>
#include <iostream>

using namespace std;

bool solution(string s)
{
    bool answer = true;
    queue<char> q;
    for (int i = 0; i < s.length(); i++) {
        if (s[i] == '(') {
            q.push(s[i]);
        } else {
            if (q.size() != 0) {
                q.pop();
            } else {
                answer = false;
                break;
            }
        }
    }
    if (q.size() != 0) 
        answer = false;
    return answer;
}

File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.java
import java.util.*;
class Solution {
    boolean solution(String s) {
        boolean answer = true;
        Stack<Character> stack = new Stack<>();
        
        stack.push(s.charAt(0));
        
        for (int i = 1; i < s.length(); i++) {
            if (stack.empty()) {
                if (s.charAt(i) == ')') {
                    answer = false;
                    break;
                }
                stack.push(s.charAt(i));
            } else {
                if (s.charAt(i) == '(') {
                    stack.push(s.charAt(i));
                } else {
                    if (stack.peek() == '(') {
                        stack.pop();
                    } else {
                        stack.push(s.charAt(i));
                    }
                }
            }
        }
        
        if (stack.size() != 0) {
            answer = false;
        }
        return answer;
    }
}